### PR TITLE
[V3 Utils] Menu system

### PR DESCRIPTION
**Pull request by palmtree5**
### Type

- [ ] Bugfix
- [x] Enhancement
- [ ] New feature

### Description of the changes
This is a possible implementation for a menu system. Basically, cog devs can pass in context, a list of either strings or embeds and a dict mapping of emoji to functions and the user will get a menu controlled by emoji. An example:

```py
from redbot.core.utils.menus import menu, DEFAULT_CONTROLS
import discord

class MyCog:
    @commands.command()
    async def mycom(self, ctx):
        pages = [discord.Embed(title="blah"), discord.Embed(title="blah 2")]
        await menu(ctx, pages, DEFAULT_CONTROLS)
```
will yield a menu where users can change pages and close the menu

`DEFAULT_CONTROLS` is basically a quick way to use the menu system if cog devs are fine using some default functions to handle menu controls (basically, these are a function to move to the next page, a function to move to the previous page, and a function to close the menu (deleting the message that is the menu))

### CODE REVIEW

File: redbot/core/utils/menus.py [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+                    controls: dict,  message: discord.Message, page: int,
+                    timeout: float, emoji: str):
+    perms = message.channel.permissions_for(ctx.guild.me)
+    if perms.manage_messages:  # Can manage messages, so remove react
+        try:
+            await message.remove_reaction(emoji, ctx.author)
+        except discord.NotFound:
+            pass
+    if page == len(pages) - 1:
+        next_page = 0  # Loop around to the first item
```
Comment (bobloy) [inline redbot/core/utils/menus.py]: Uses `next_page` as variable and `next_page` and function

**Comment by bobloy** [comment]
It appears multiple instances of a menu will trigger each other's button presses. 
i.e ❌ will close all menus

**Comment by Bakersbakebread** [comment]
Upvote

File: redbot/core/utils/menus.py [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+        for key in controls.keys():
+            await message.add_reaction(key)
+    else:
+        if isinstance(current_page, discord.Embed):
+            await message.edit(embed=current_page)
+        else:
+            await message.edit(content=current_page)
+
+    def react_check(r, u):
+        return u == ctx.author and str(r.emoji) in controls.keys()
```
Comment (bobloy) [inline redbot/core/utils/menus.py]: add `and r.message==message`

File: redbot/core/utils/menus.py [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+                     controls: dict,  message: discord.Message, page: int,
+                     timeout: float, emoji: str):
+    if message:
+        await message.delete()
+    return None
+
+
+DEFAULT_CONTROLS = {
+    "➡": next_page,
+    "⬅": prev_page,
```
Comment (bobloy) [inline redbot/core/utils/menus.py]: Swap order of `next_page` and `prev_page` for clearer buttons

**Review by bobloy — CHANGES_REQUESTED** [review]

**Review by Kowlin — APPROVED** [review]

**Pull request merged**